File: skills/skill-reviewer/scripts/start_skill_dashboard.py
#!/usr/bin/env python3
"""Project one Skill Reviewer run and start its temporary local control plane."""

from __future__ import annotations

import argparse
import errno
import hashlib
import json
import secrets
import signal
import sys
import webbrowser
from http.server import ThreadingHTTPServer
from pathlib import Path
from threading import Event, Thread
from urllib.parse import urlencode

from dashboard_bundle import (
    DashboardBundleError,
    MaterializedDashboardUi,
    materialize_dashboard_ui,
)
from serve_skill_dashboard import (
    DashboardServerError,
    _validate_loopback_bind_host,
    create_handler,
    validate_sources,
)
from skill_eval_runtime import ManifestError, project_dashboard


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--workspace",
        type=Path,
        required=True,
        help="Locked eval workspace containing execution-plan.json.",
    )
    parser.add_argument(
        "--state",
        type=Path,
        help="Optional evolution-state.json outside the run workspace.",
    )
    parser.add_argument(
        "--task-root",
        type=Path,
        help="Append-only lead-Agent task ledger outside the evidence workspace.",
    )
    parser.add_argument(
        "--ui-dir",
        type=Path,
        help=(
            "Trusted local Dashboard build for development or offline use. "
            "When omitted, the launcher anonymously downloads the pinned bundle "
            "into a temporary directory and removes it on exit."
        ),
    )
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument(
        "--port",
        type=int,
        default=8765,
        help="Preferred local port. Use 0 for an operating-system-assigned port.",
    )
    parser.add_argument(
        "--port-attempts",
        type=int,
        default=3,
        help="Number of consecutive ports to try without stopping another process.",
    )
    parser.add_argument(
        "--refresh-seconds",
        type=float,
        default=3.0,
        help=(
            "Reproject the run at this interval while serving so new execution "
            "evidence appears in the same page. Use 0 to disable."
        ),
    )
    parser.add_argument(
        "--open",
        action="store_true",
        help="Open the local control-plane URL after the server starts.",
    )
    parser.add_argument(
        "--serve-existing",
        action="store_true",
        help="Serve an existing dashboard-data.json without reprojecting the run.",
    )
    parser.add_argument(
        "--prepare-only",
        action="store_true",
        help=(
            "Project and validate the read model, then exit without downloading "
            "or starting the optional control plane."
        ),
    )
    parser.add_argument(
        "--user-approved-control-plane",
        action="store_true",
        help=(
            "Assert that the current request or a structured consent question "
            "explicitly authorized this temporary local control-plane session. "
            "Required unless --prepare-only is used."
        ),
    )
    args = parser.parse_args(argv)
    if args.port < 0 or args.port > 65535:
        parser.error("--port must be between 0 and 65535")
    if args.port_attempts < 1 or args.port_attempts > 20:
        parser.error("--port-attempts must be between 1 and 20")
    if args.refresh_seconds < 0 or args.refresh_seconds > 3600:
        parser.error("--refresh-seconds must be between 0 and 3600")
    if args.port > 0 and args.port + args.port_attempts - 1 > 65535:
        parser.error("the requested port range exceeds 65535")
    if not args.prepare_only and not args.user_approved_control_plane:
        parser.error(
            "starting the optional Dashboard requires explicit user approval; "
            "use an existing explicit Dashboard request or ask once with a "
            "standalone structured question, then pass "
            "--user-approved-control-plane only after an affirmative answer"
        )
    return args


def _port_candidates(preferred: int, attempts: int) -> list[int]:
    if preferred == 0:
        return [0]
    return list(range(preferred, preferred + attempts))


def _bind_server(
    *, host: str, preferred_port: int, attempts: int, handler: type
) -> ThreadingHTTPServer:
    last_error: OSError | None = None
    for port in _port_candidates(preferred_port, attempts):
        try:
            return ThreadingHTTPServer((host, port), handler)
        except OSError as error:
            last_error = error
            if error.errno not in {errno.EADDRINUSE, errno.EACCES}:
                raise
    assert last_error is not None
    raise DashboardServerError(
        "no dashboard port is available in the requested range; "
        "choose another --port or use --port 0"
    ) from last_error


def _control_plane_origin(host: str, port: int) -> str:
    authority = f"[{host}]" if ":" in host else host
    return f"http://{authority}:{port}"


def _control_plane_url(origin: str, session_token: str) -> str:
    return f"{origin}/skill-reviewer/#{urlencode({'session': session_token})}"


def prepare_dashboard(args: argparse.Namespace) -> dict[str, object]:
    workspace = args.workspace.resolve()
    if not args.serve_existing:
        project_dashboard(
            workspace=workspace,
            output=workspace / "dashboard-data.json",
            state_path=args.state,
        )
    report = validate_sources(workspace, args.task_root)
    return {
        **report,
        "dashboard_hosted": False,
        "control_plane_started": False,
        "projected": not args.serve_existing,
        "projection_source": (
            "fresh_run_projection" if not args.serve_existing else "existing_projection"
        ),
        "evidence_uploaded": False,
    }


def _projection_digest(workspace: Path) -> str:
    return hashlib.sha256((workspace / "dashboard-data.json").read_bytes()).hexdigest()


def _watch_projection(
    *,
    workspace: Path,
    state_path: Path | None,
    interval: float,
    initial_digest: str,
    stop: Event,
) -> None:
    """Keep one active run observable without replacing its evidence source."""
    previous_digest = initial_digest
    previous_error: str | None = None
    while not stop.wait(interval):
        try:
            data = project_dashboard(
                workspace=workspace,
                output=workspace / "dashboard-data.json",
                state_path=state_path,
            )
            current_digest = _projection_digest(workspace)
            if current_digest != previous_digest or previous_error is not None:
                run = data.get("run")
                print(
                    json.dumps(
                        {
                            "event": "dashboard_projection_updated",
                            "run_id": run.get("id") if isinstance(run, dict) else None,
                            "status": run.get("status") if isinstance(run, dict) else None,
                            "verification_level": (
                                run.get("verification_level")
                                if isinstance(run, dict)
                                else None
                            ),
                        },
                        ensure_ascii=False,
                    ),
                    file=sys.stderr,
                    flush=True,
                )
            previous_digest = current_digest
            previous_error = None
        except (ManifestError, OSError, ValueError) as error:
            message = str(error)
            if message != previous_error:
                print(
                    json.dumps(
                        {
                            "event": "dashboard_projection_failed",
                            "error": message,
                            "retrying": True,
                        },
                        ensure_ascii=False,
                    ),
                    file=sys.stderr,
                    flush=True,
                )
            previous_error = message


def _ui_report(ui: MaterializedDashboardUi) -> dict[str, object]:
    return {
        "ui_mode": ui.mode,
        "ui_downloaded": ui.temporary,
        "ui_temporary": ui.temporary,
        "ui_removed_on_exit": ui.temporary,
        "ui_integrity_verified": ui.integrity_verified,
        "ui_archive_sha256": ui.archive_sha256,
        "ui_tree_sha256": ui.tree_sha256,
        "ui_download_authenticated": False,
        "github_token_used": False,
    }


def _raise_keyboard_interrupt(_signum: int, _frame: object) -> None:
    raise KeyboardInterrupt


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv or sys.argv[1:])
    try:
        _validate_loopback_bind_host(args.host)
        report = prepare_dashboard(args)
        if args.prepare_only:
            print(
                json.dumps(
                    {
                        **report,
                        "ui_downloaded": False,
                        "ui_removed_on_exit": False,
                    },
                    ensure_ascii=False,
                    indent=2,
                )
            )
            return 0

        temporary_ui_removed = False
        with materialize_dashboard_ui(args.ui_dir) as ui:
            server: ThreadingHTTPServer | None = None
            projection_stop = Event()
            projection_thread: Thread | None = None
            session_token = secrets.token_urlsafe(32)
            try:
                handler = create_handler(
                    args.workspace,
                    args.task_root,
                    session_token=session_token,
                    static_ui_root=ui.root,
                )
                server = _bind_server(
                    host=args.host,
                    preferred_port=args.port,
                    attempts=args.port_attempts,
                    handler=handler,
                )
                bound_host, bound_port = server.server_address[:2]
                origin = _control_plane_origin(str(bound_host), int(bound_port))
                url = _control_plane_url(origin, session_token)
                launch_report = {
                    **report,
                    **_ui_report(ui),
                    "user_approved_control_plane": True,
                    "control_plane_started": True,
                    "base_url": origin,
                    "data_url": f"{origin}/dashboard-data.json",
                    "session_url": f"{origin}/dashboard-session.json",
                    "url": url,
                    "port": bound_port,
                    "projection_mode": (
                        "watching"
                        if not args.serve_existing and args.refresh_seconds > 0
                        else "static"
                    ),
                    "refresh_seconds": (
                        args.refresh_seconds
                        if not args.serve_existing and args.refresh_seconds > 0
                        else 0
                    ),
                    "dashboard_session": {
                        "contract": "skill-reviewer.dashboard-session",
                        "run_id": report.get("run_id"),
                        "page_url": url,
                        "local_origin": origin,
                        "owner": "lead_agent",
                        "lifecycle": "temporary-local-control-plane",
                        "evidence_transport": "same-origin-loopback-only",
                        "evidence_uploaded": False,
                        "capability_transport": "url-fragment-to-request-header",
                        "ui_integrity_verified": ui.integrity_verified,
                        "ui_downloaded": ui.temporary,
                        "ui_removed_on_exit": ui.temporary,
                        "browser_executes_actions": False,
                        "agent_handoff": {
                            "contract": "skill-reviewer.dashboard-agent-handoff",
                            "mode": "durable_local_ledger",
                            "agent_session_state": "unbound",
                            "can_wake_agent_session": False,
                            "persists_after_agent_session_end": True,
                            "task_root": report.get("task_root"),
                        },
                    },
                }
                print(json.dumps(launch_report, ensure_ascii=False), flush=True)
                if args.open and not webbrowser.open(url):
                    print(
                        "dashboard browser could not be opened automatically; "
                        "use the printed local URL",
                        file=sys.stderr,
                    )
                if not args.serve_existing and args.refresh_seconds > 0:
                    projection_thread = Thread(
                        target=_watch_projection,
                        kwargs={
                            "workspace": args.workspace.resolve(),
                            "state_path": args.state,
                            "interval": args.refresh_seconds,
                            "initial_digest": _projection_digest(
                                args.workspace.resolve()
                            ),
                            "stop": projection_stop,
                        },
                        name="skill-reviewer-dashboard-projector",
                        daemon=True,
                    )
                    projection_thread.start()
                if hasattr(signal, "SIGTERM"):
                    signal.signal(signal.SIGTERM, _raise_keyboard_interrupt)
                server.serve_forever()
            except KeyboardInterrupt:
                pass
            finally:
                projection_stop.set()
                if projection_thread is not None:
                    projection_thread.join(timeout=2)
                if server is not None:
                    server.server_close()
            temporary_ui_removed = ui.temporary
        if temporary_ui_removed:
            print(
                json.dumps(
                    {
                        "event": "dashboard_temporary_ui_removed",
                        "evidence_uploaded": False,
                    },
                    ensure_ascii=False,
                ),
                file=sys.stderr,
                flush=True,
            )
        return 0
    except (
        DashboardBundleError,
        DashboardServerError,
        ManifestError,
        OSError,
        ValueError,
    ) as error:
        print(json.dumps({"ok": False, "error": str(error)}, ensure_ascii=False))
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/skill-reviewer/scripts/start_skill_dashboard.mjs
#!/usr/bin/env node

/** Project one Skill Reviewer run and start its temporary local Dashboard session. */

import { createHash } from "node:crypto";
import { spawn } from "node:child_process";
import { readFileSync, realpathSync } from "node:fs";
import { join, resolve } from "node:path";

import {
  DashboardBundleError,
  materializeDashboardUi,
} from "./dashboard_bundle.mjs";
import {
  DASHBOARD_LAUNCH_SESSION_CONTRACT,
  ManifestError,
} from "./lib/skill-eval-contracts.mjs";
import { projectDashboard } from "./lib/skill-eval-dashboard.mjs";
import { isMainModule } from "./lib/module-entrypoint.mjs";
import {
  DashboardServerError,
  createDashboardServer,
  randomSessionToken,
  validateLoopbackBindHost,
  validateSources,
} from "./serve_skill_dashboard.mjs";

class DashboardLauncherUsageError extends Error {
  constructor(message) {
    super(message);
    this.name = "DashboardLauncherUsageError";
  }
}

function displayJson(value) {
  if (Array.isArray(value)) return `[${value.map(displayJson).join(", ")}]`;
  if (value !== null && typeof value === "object") {
    return `{${Object.entries(value).map(([key, item]) => `${JSON.stringify(key)}: ${displayJson(item)}`).join(", ")}}`;
  }
  return JSON.stringify(value);
}

function numberOption(raw, option) {
  const value = Number(raw);
  if (!Number.isFinite(value)) throw new DashboardLauncherUsageError(`${option} must be a number`);
  return value;
}

export function parseArgs(argv) {
  const values = {
    host: "127.0.0.1",
    port: 8765,
    portAttempts: 3,
    refreshSeconds: 3,
    open: false,
    serveExisting: false,
    prepareOnly: false,
    userApprovedDashboard: false,
    state: null,
    uiDir: null,
  };
  if (argv.includes("--help") || argv.includes("-h")) return { help: true };
  const flags = new Map([
    ["--open", "open"],
    ["--serve-existing", "serveExisting"],
    ["--prepare-only", "prepareOnly"],
    ["--user-approved-dashboard", "userApprovedDashboard"],
  ]);
  const options = new Set([
    "--workspace", "--state", "--ui-dir", "--host", "--port",
    "--port-attempts", "--refresh-seconds",
  ]);
  const seen = new Set();
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (flags.has(token)) {
      if (seen.has(token)) throw new DashboardLauncherUsageError(`${token} may be provided only once`);
      seen.add(token);
      values[flags.get(token)] = true;
      continue;
    }
    if (!options.has(token)) throw new DashboardLauncherUsageError(`unknown option: ${token}`);
    if (seen.has(token)) throw new DashboardLauncherUsageError(`${token} may be provided only once`);
    seen.add(token);
    const value = argv[++index];
    if (value === undefined) throw new DashboardLauncherUsageError(`${token} requires a value`);
    if (token === "--workspace") values.workspace = resolve(value);
    else if (token === "--state") values.state = resolve(value);
    else if (token === "--ui-dir") values.uiDir = resolve(value);
    else if (token === "--host") values.host = value;
    else if (token === "--port") values.port = numberOption(value, token);
    else if (token === "--port-attempts") values.portAttempts = numberOption(value, token);
    else values.refreshSeconds = numberOption(value, token);
  }
  if (!values.workspace) throw new DashboardLauncherUsageError("--workspace is required");
  if (!Number.isInteger(values.port) || values.port < 0 || values.port > 65535) {
    throw new DashboardLauncherUsageError("--port must be between 0 and 65535");
  }
  if (!Number.isInteger(values.portAttempts) || values.portAttempts < 1 || values.portAttempts > 20) {
    throw new DashboardLauncherUsageError("--port-attempts must be between 1 and 20");
  }
  if (values.refreshSeconds < 0 || values.refreshSeconds > 3600) {
    throw new DashboardLauncherUsageError("--refresh-seconds must be between 0 and 3600");
  }
  if (values.port > 0 && values.port + values.portAttempts - 1 > 65535) {
    throw new DashboardLauncherUsageError("the requested port range exceeds 65535");
  }
  if (!values.prepareOnly && !values.userApprovedDashboard) {
    throw new DashboardLauncherUsageError(
      "starting the optional Dashboard requires an explicit user request; pass --user-approved-dashboard only after that request",
    );
  }
  return values;
}

function usage() {
  return [
    "Usage: start_skill_dashboard.mjs --workspace PATH [options]",
    "Options:",
    "  --state PATH --ui-dir PATH --host HOST --port PORT",
    "  --port-attempts N --refresh-seconds N --open --serve-existing",
    "  --prepare-only --user-approved-dashboard",
  ].join("\n");
}

export function portCandidates(preferred, attempts) {
  if (preferred === 0) return [0];
  return Array.from({ length: attempts }, (_, index) => preferred + index);
}

function listen(server, host, port) {
  return new Promise((resolveListen, rejectListen) => {
    const onError = (error) => rejectListen(error);
    server.once("error", onError);
    server.listen(port, host, () => {
      server.off("error", onError);
      resolveListen();
    });
  });
}

export async function bindDashboardServer({
  host,
  preferredPort,
  attempts,
  workspace,
  sessionToken,
  staticUiRoot,
}) {
  let lastError = null;
  for (const port of portCandidates(preferredPort, attempts)) {
    const server = createDashboardServer({ workspace, sessionToken, staticUiRoot });
    try {
      await listen(server, host, port);
      return server;
    } catch (error) {
      lastError = error;
      if (!["EADDRINUSE", "EACCES"].includes(error?.code)) throw error;
    }
  }
  throw new DashboardServerError(
    "no dashboard port is available in the requested range; choose another --port or use --port 0",
    { cause: lastError },
  );
}

export function dashboardOrigin(host, port) {
  const authority = host.includes(":") ? `[${host}]` : host;
  return `http://${authority}:${port}`;
}

export function dashboardUrl(origin, sessionToken) {
  return `${origin}/skill-reviewer/#${new URLSearchParams({ session: sessionToken })}`;
}

export async function prepareDashboard(args) {
  const workspace = realpathSync(resolve(args.workspace));
  if (!args.serveExisting) {
    await projectDashboard({
      workspace,
      output: join(workspace, "dashboard-data.json"),
      statePath: args.state ?? undefined,
    });
  }
  const report = validateSources(workspace);
  return {
    ...report,
    dashboard_hosted: false,
    dashboard_session_started: false,
    projected: !args.serveExisting,
    projection_source: args.serveExisting ? "existing_projection" : "fresh_run_projection",
    evidence_uploaded: false,
  };
}

function projectionDigest(workspace) {
  return createHash("sha256")
    .update(readFileSync(join(workspace, "dashboard-data.json")))
    .digest("hex");
}

export function watchProjection({ workspace, statePath, intervalSeconds, initialDigest }) {
  let stopped = false;
  let timer = null;
  let previousDigest = initialDigest;
  let previousError = null;
  let active = Promise.resolve();

  async function update() {
    if (stopped) return;
    try {
      const data = await projectDashboard({
        workspace,
        output: join(workspace, "dashboard-data.json"),
        statePath: statePath ?? undefined,
      });
      const currentDigest = projectionDigest(workspace);
      if (currentDigest !== previousDigest || previousError !== null) {
        const run = data && typeof data === "object" && !Array.isArray(data) ? data.run : null;
        process.stderr.write(`${displayJson({
          event: "dashboard_projection_updated",
          run_id: run && typeof run === "object" ? run.id ?? null : null,
          status: run && typeof run === "object" ? run.status ?? null : null,
          verification_level: run && typeof run === "object" ? run.verification_level ?? null : null,
        })}\n`);
      }
      previousDigest = currentDigest;
      previousError = null;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      if (message !== previousError) {
        process.stderr.write(`${displayJson({
          event: "dashboard_projection_failed",
          error: message,
          retrying: true,
        })}\n`);
      }
      previousError = message;
    } finally {
      if (!stopped) {
        timer = setTimeout(() => {
          active = update();
        }, intervalSeconds * 1000);
        timer.unref?.();
      }
    }
  }

  timer = setTimeout(() => {
    active = update();
  }, intervalSeconds * 1000);
  timer.unref?.();

  return {
    async stop() {
      stopped = true;
      if (timer) clearTimeout(timer);
      await active;
    },
  };
}

function uiReport(ui) {
  return {
    ui_mode: ui.mode,
    ui_downloaded: ui.temporary,
    ui_temporary: ui.temporary,
    ui_removed_on_exit: ui.temporary,
    ui_integrity_verified: ui.integrity_verified,
    ui_archive_sha256: ui.archive_sha256,
    ui_tree_sha256: ui.tree_sha256,
    ui_download_authenticated: false,
    github_token_used: false,
  };
}

function openBrowser(url) {
  let command;
  let args;
  if (process.platform === "darwin") {
    command = "open";
    args = [url];
  } else if (process.platform === "win32") {
    command = "cmd";
    args = ["/c", "start", "", url];
  } else {
    command = "xdg-open";
    args = [url];
  }
  try {
    const child = spawn(command, args, { detached: true, stdio: "ignore" });
    child.unref();
    return true;
  } catch {
    return false;
  }
}

function waitForStop(server) {
  return new Promise((resolveStop) => {
    const stop = () => resolveStop();
    process.once("SIGINT", stop);
    process.once("SIGTERM", stop);
    server.once("close", stop);
  });
}

function closeServer(server) {
  if (!server?.listening) return Promise.resolve();
  return new Promise((resolveClose) => server.close(() => resolveClose()));
}

export async function main(argv = process.argv.slice(2)) {
  let args;
  try {
    args = parseArgs(argv);
  } catch (error) {
    process.stderr.write(`start_skill_dashboard.mjs: error: ${error.message}\n`);
    return 2;
  }
  if (args.help) {
    process.stdout.write(`${usage()}\n`);
    return 0;
  }

  let materialized = null;
  let server = null;
  let watcher = null;
  let temporaryUiRemoved = false;
  try {
    validateLoopbackBindHost(args.host);
    const report = await prepareDashboard(args);
    if (args.prepareOnly) {
      process.stdout.write(`${JSON.stringify({
        ...report,
        ui_downloaded: false,
        ui_removed_on_exit: false,
      }, null, 2)}\n`);
      return 0;
    }

    materialized = await materializeDashboardUi({ uiDir: args.uiDir });
    const ui = materialized.ui;
    const sessionToken = randomSessionToken();
    server = await bindDashboardServer({
      host: args.host,
      preferredPort: args.port,
      attempts: args.portAttempts,
      workspace: args.workspace,
      sessionToken,
      staticUiRoot: ui.root,
    });
    const address = server.address();
    const origin = dashboardOrigin(String(address.address), Number(address.port));
    const url = dashboardUrl(origin, sessionToken);
    const watching = !args.serveExisting && args.refreshSeconds > 0;
    const launchReport = {
      ...report,
      ...uiReport(ui),
      user_approved_dashboard: true,
      dashboard_session_started: true,
      base_url: origin,
      data_url: `${origin}/dashboard-data.json`,
      session_url: `${origin}/dashboard-session.json`,
      url,
      port: address.port,
      projection_mode: watching ? "watching" : "static",
      refresh_seconds: watching ? args.refreshSeconds : 0,
      dashboard_session: {
        contract: DASHBOARD_LAUNCH_SESSION_CONTRACT,
        run_id: report.run_id ?? null,
        page_url: url,
        local_origin: origin,
        owner: "lead_agent",
        lifecycle: "temporary-local-dashboard",
        evidence_transport: "same-origin-loopback-only",
        evidence_uploaded: false,
        capability_transport: "url-fragment-to-request-header",
        ui_integrity_verified: ui.integrity_verified,
        ui_downloaded: ui.temporary,
        ui_removed_on_exit: ui.temporary,
      },
    };
    process.stdout.write(`${JSON.stringify(launchReport)}\n`);
    if (args.open && !openBrowser(url)) {
      process.stderr.write("dashboard browser could not be opened automatically; use the printed local URL\n");
    }
    if (watching) {
      watcher = watchProjection({
        workspace: realpathSync(resolve(args.workspace)),
        statePath: args.state,
        intervalSeconds: args.refreshSeconds,
        initialDigest: projectionDigest(realpathSync(resolve(args.workspace))),
      });
    }
    await waitForStop(server);
    return 0;
  } catch (error) {
    if (
      error instanceof DashboardBundleError || error instanceof DashboardServerError ||
      error instanceof ManifestError || error instanceof Error
    ) {
      process.stdout.write(`${JSON.stringify({ ok: false, error: error.message })}\n`);
      return 2;
    }
    process.stdout.write(`${JSON.stringify({ ok: false, error: String(error) })}\n`);
    return 2;
  } finally {
    await watcher?.stop();
    await closeServer(server);
    if (materialized) {
      temporaryUiRemoved = materialized.ui.temporary;
      materialized.dispose();
    }
    if (temporaryUiRemoved) {
      process.stderr.write(`${displayJson({
        event: "dashboard_temporary_ui_removed",
        evidence_uploaded: false,
      })}\n`);
    }
  }
}

if (isMainModule(import.meta.url)) process.exitCode = await main();
